rllib/connectors/learner/add_one_ts_to_episodes_and_truncate.py

```diff
@@ -101,10 +101,24 @@ def __call__(
         # batch: - - - - - - - T B0- - - - - R Bx- - - - R Bx
         # mask : t t t t t t t t f t t t t t t f t t t t t f
 
-        shared_data["_sa_episodes_lengths"] = {}
-        for sa_episode in self.single_agent_episode_iterator(
-            episodes, agents_that_stepped_only=False
+        for i, sa_episode in enumerate(
+            self.single_agent_episode_iterator(episodes, agents_that_stepped_only=False)
         ):
+            # TODO (sven): This is a little bit of a hack: By expanding the Episode's
+            #  ID, we make sure that each episode chunk in `episodes` is treated as a
+            #  separate episode in the `self.add_n_batch_items` below. Some algos (e.g.
+            #  APPO) may have >1 episode chunks from the same episode (same ID) in the
+            #  training data, thus leading to a malformatted batch in case of
+            #  RNN-triggered zero-padding of the train batch.
+            #  For example, if e1 (id=a len=4) and e2 (id=a len=5) are two chunks of the
+            #  same episode in `episodes`, the resulting batch would have an additional
+            #  timestep in the middle of the episode's "row":
+            #  {  "obs": {
```

Comment:
Couldn't we just concat the episode and then cut them into sequence len pieces?

Reply:
Yeah, I was thinking about that, too. But I think this would cost too much extra time? It would certainly be of advantage b/c it would eliminate the "bad" samples, where we have very short episode chunks (a new episode started close to the end (e.g. 2 timesteps away) of the EnvRunner's rollout fragment length) and then zero-pad that trajectory all the way to the end, where instead we could concatenate that short piece with the longer successor piece to give it "more meaning".

We could try adding this logic to this same connector (for now) and run some experiments with CartPole (generally short episodes) and Pong (generally quite long episodes) to see, whether there is a performance delta ...